sdk/storage/AzureStorageBlob/Source/DataStructures/StorageStreamUploader.swift

```diff
@@ -121,7 +121,8 @@ internal class ChunkUploader {
             let tempData = fileHandle.readData(ofLength: chunkSize)
             buffer.append(tempData)
         } catch {
-            completion(.failure(error), HTTPResponse(request: nil, statusCode: nil))
+            let request = try? HTTPRequest(method: .put, url: url, headers: HTTPHeaders())
```

Comment:
If we're synthesizing a fake response here just to make the completion handler happy, maybe we should rethink the design of the handler. Do we need to supply a response at all?

Reply:
It's a tradeoff between SDK code and client code. If I did make it `HTTPResponse?`, then the client code is going to have to add guard statements or otherwise unwrap that optional every time they want to use it. I could make it `HTTPResponse!` but (assuming XCode or a linter don't howl in anger) the app would crash if the user tried to use the response that was nil.

Thus I settled on preferring to always return some kind of Response object, even if it is junk. Ugly code inside the SDK, but I figure I'd rather have it there.

That was my rationale, but I'm happy to examine other options!

Reply:
Ahh yes, I see - the problem here is encoded in the type. The assumption is that regardless of whether the operation succeeds or fails, there will always be a HTTPResponse available. I don't think that's a valid assumption, so I would be in favor of changing the type signature for ResultHandler to make the response optional. It makes perfect sense that there will sometimes be errors for which no response (or even no request) object is available. For comparison, OkHttp has a `Callback` interface with an `onResponse(Request, Response)` and an `onFailure(Request, Exception)` so they seem to at least make the assumption that there will always be a request available.